ably/realtime_client.go

```diff
@@ -72,37 +72,42 @@ func (c *RealtimeClient) onReconnectMsg(msg *proto.ProtocolMessage) {
 		if c.Connection.id == msg.ConnectionID {
 			if msg.Error != nil {
 				// (RTN15c2)
+
+				// I'm a bit confused here as the spec says to set CONNECTED event but
+				// there is no event mechanism only state changes and we are not supposed to
+				// change the state here.
 				c.Connection.state.Lock()
 				c.Connection.setState(StateConnConnected, msg.Error)
-
-				// According to the spec we need to set Connection#errorReason but Conn
-				// doesn't have errorReason field.However it has Reason method that
-				// returns the last know error which in our case will be be msg.Error so
-				// it will be conforming to spec.
-				//
-				// Adding Conn.errorReason won't work in our case as it will be private
-				// according to Go conventions. So, I'm choosing to only set state with
-				// appropriate reason and assuming users will use Conn.Reason to access
-				// the error.
-				//
-				// TODO: Add Conn.ErrorReason  field.
 				c.Connection.state.Unlock()
+				c.Connection.queue.Flush()
 				return
 			}
 			// (RTN15c1)
+			c.Connection.queue.Flush()
```

Comment:
Since this is also what happens on a normal transition to CONNECTED, and there's also some other common logic between first-time connection and reconnection, I'm worried we're duplicating too much. Maybe we should refactor so that reconnection includes that common logic?